cmd/dagger/client.go

```diff
@@ -14,3 +43,183 @@ var clientCmd = &cobra.Command{
+					Dev: dev,
+				}).
+				GeneratedContextDirectory().
+				Export(ctx, contextDirPath)
+			if err != nil {
+				return fmt.Errorf("failed to export client: %w", err)
+			}
+
+			w := cmd.OutOrStdout()
+			fmt.Fprintf(w, "Generated client at %s\n", outputPath)
+
+			return nil
+		})
+	},
+	Annotations: map[string]string{
+		"experimental": "true",
+	},
+}
+
+var clientUninstallCmd = &cobra.Command{
+	Use:     "uninstall [path]",
+	Short:   "Remove a Dagger client from the module",
+	Example: "dagger client uninstall ./dagger",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return withEngine(cmd.Context(), client.Params{}, func(ctx context.Context, engineClient *client.Client) error {
+			if len(args) != 1 {
+				return fmt.Errorf("expected only the path to the generated client to be set as argument")
+			}
+
+			cwd, err := pathutil.Getwd()
+			if err != nil {
+				return fmt.Errorf("failed to get current working directory: %w", err)
+			}
+
+			path := args[0]
+
+			if filepath.IsAbs(path) {
+				path, err = filepath.Rel(cwd, path)
+				if err != nil {
+					return fmt.Errorf("failed to get relative path: %w", err)
+				}
+			}
+
+			dag := engineClient.Dagger()
+
+			mod, err := initializeClientGeneratorModule(ctx, dag, ".")
+			if err != nil {
+				return fmt.Errorf("failed to initialize client generator module: %w", err)
+			}
+
+			contextDirPath, err := mod.Source.LocalContextDirectoryPath(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to get local context directory path: %w", err)
+			}
+
+			_, err = mod.Source.
+				WithoutClient(path).
+				GeneratedContextDirectory().
+				Export(ctx, contextDirPath)
+			if err != nil {
+				return fmt.Errorf("failed to remove client from module: %w", err)
+			}
+
+			w := cmd.OutOrStdout()
+			fmt.Fprintf(w,
+				"Client at %s removed from config.\n"+
+					"Please manually remove any remaining files of the generated client from your host\n",
+				path)
+
+			return nil
+		})
+	},
+}
+
+//go:embed clientconf.graphql
+var loadModClientConfQuery string
+
+var clientListCmd = &cobra.Command{
+	Use:     "list",
+	Short:   "List all Dagger clients installed in the current module",
+	Example: "dagger client list",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return withEngine(cmd.Context(), client.Params{}, func(ctx context.Context, engineClient *client.Client) error {
+			dag := engineClient.Dagger()
+
+			mod, err := initializeClientGeneratorModule(ctx, dag, ".")
+			if err != nil {
+				return fmt.Errorf("failed to initialize client generator module: %w", err)
+			}
+
+			moduleSourceID, err := mod.Source.ID(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to list clients: failed to get module source id: %w", err)
+			}
+
+			var res struct {
+				Source struct {
+					ConfigClients []struct {
+						Generator string
+						Directory string
+					}
+				}
+			}
+
+			err = dag.Do(ctx, &dagger.Request{
+				Query: loadModClientConfQuery,
```

Comment:
Why not just use the code-gened API directly?

Reply:
Because it's complex with the codegened API to get nested fields in an array so I used that "hack" to make it cleaner on the code side :)